concepts/random/about.md

```diff
@@ -0,0 +1,171 @@
+# About
+
+Many programs need (apparently) random values to simulate real-world events.
+
+Common, familiar examples include:
+- A coin toss: a random value from `('H', 'T')`.
+- The roll of a die: a random integer from 1 to 6.
+- Shuffling a deck of cards: a random ordering of a card list.
+
+Generating truly random values with a computer is a surprisingly difficult technical challenge, so you may see these results referred to as "pseudorandom".
+
+In practice, a well-designed library like the [`random`][random] module in the Python standard library is fast, flexible, and gives results that are amply good enough for most applications in modelling, simulation and games.
+
+The rest of this page will list a few of the most common functions in `random`.
+
+We encourage you to explore the full [`random`][random] documentation, as there are many more options.
+
+## Important Warning!
+
+The `random` module should __NOT__ be used for security and cryptographic applications.
+
+Instead, Python provides the [`secrets`][secrets] module.
+This is specially optimized for cryptographic security.
+
+## Create random integers
+
+The `randrange()` function has three forms, to select a random value from `range(start, stop, step)`:
+- `randrange(stop)` gives an integer `n` such that `0 <= n < stop`
+- `randrange(start, stop)` gives an integer `n` such that `start <= n < stop`
+- `randrange(start, stop, step)` gives an integer `n` such that `start <= n < stop` and `n` is in the sequence `start, start + step, start + 2*step...`
+
+For the common case where `step == 1`, the `randint(a, b)` function may be more convenient and readable.
+
+Possible results from `randint()` _include_ the upper bound, so `randint(a, b)` is the same as `randrange(a, b+1)`.
+
+```python
+>>> import random
+
+>>> random.randrange(500)
+219
+>>> [random.randrange(0, 10, 2) for _ in range(10)]
+[2, 8, 4, 0, 4, 2, 6, 6, 8, 8]
+
+>>> random.randint(1, 6)  # roll a die
+4
+```

```

Comment:
A few minutes later, with a bit more thought: this would be a killer error in `introduction.md`. For `about.md` I think we have more leeway to forward-reference concepts the students haven't reached yet. May need an explanatory comment.

Reply:
> Another thing I worry about: these are high up but absolutely depend on import. How do we resolve that?

Errr.  Yeah.  That said, `import` is really easy to show by example.  At least that's what I tell myself every time it comes up.  But if it is also raising concerns for you, we might want to brainstorm on how we address that.

One option, of course, is to move concepts around.  I like the idea of keeping math related stuff in a cluster, but I also don't want to send anyone screaming for the exit.  Now, none of these are on the "critical path" of prerequisites, but a lot of folx might very well go through them because they are _there_.

Option two would be to briefly show & explain how you do an `import`, and hope that we don't get complaints about the hand-waving.  I am not sure we would, since its really straightforward.  Something along the lines of:

     To use the `random` module, you must first import it like so:

       ```python
       import random

      random.choice([1,2,3,4,5,6,7])
      ```

Option three would be to write up an `import` concept.  I am allergic to this for several reasons:
  1.  Doing it early means skipping the (important) details around namespacing, names, and aliases.
  2.  If we leave out namespacing, names, and aliases -- what is left to include that _isn't_ just an example?
  3.  Even if we do a really good job, how much of the detail will help students, and how much will they retain?


I think I might be in favor of option two.  What are your thoughts?

Reply:
> I think I might be in favor of option two

Agreed, with the caveat that we may also need `from xxx import yyy`. I've used that a lot in `datetime`, though not in the numbers cluster.

Next step: when do we discourage `from xxx import *`. Presumably, after namespaces.

Reply:
Oh, I might say immediately.  But certainly after namespaces.